File: AppointmentService.java
package edu.snhu.badamson.appointment;

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;
import lombok.EqualsAndHashCode;

@EqualsAndHashCode
public class AppointmentService {
	private List<Appointment> appointments;
	
	public AppointmentService() {
		appointments = new ArrayList<>();			
	}

	/**
	 * Adds a new appointment to the service
	 * Checks that the ID is unique
	 * @param newAppointment is the appointment to add
	 * @throws IllegalArgumentException if the appointment is null or a duplicate ID.
	 */
	public void add(Appointment newAppointment) {
		 if (newAppointment == null) {
	            throw new IllegalArgumentException("Appointment cannot be null.");
		 }
		
		boolean isDuplicate = false;
		for (Appointment appointment : appointments) {
			if (appointment.getId().equals(newAppointment.getId())) {
				isDuplicate = true;
				break;
			}
		}
		
		if (isDuplicate) {
			throw new IllegalArgumentException("Cannot add appointment with duplicate ID: " + newAppointment.getId());
		}
		appointments.add(newAppointment);		
	}
	
	/**
	 * Deletes an appointment by its ID.
	 * @param id The ID of the appointment to be deleted.
	 * @throws IllegalArgumentException if the ID is null.
	 * @throws NoSuchElementException if no appointment with the given ID exists.
	 */
	public void delete(String id) {
	    if (id == null) {
	        throw new IllegalArgumentException("ID cannot be null.");
	    }
	    
	    Appointment appointmentToDelete = null;
	    
	    for (Appointment appointment : appointments) {
	        if (appointment.getId().equals(id)) {
	        	appointmentToDelete = appointment;
	            break;
	        }
	    }
	    
	    if (appointmentToDelete == null) {
	        throw new NoSuchElementException("Appointment with ID " + id + " does not exist.");
	    }
	    
	    appointments.remove(appointmentToDelete);
	}
	
	/**
	 * Get the appointment by ID.
	 * Returns the appointment if found, otherwise returns null.
	 * @param id - the appointment ID.
	 * @return the appointment or null if ID does not exist.
	 * @throws IllegalArgumentExcpetion if the ID is null.
	 */
	public Appointment get(String id) {
		if (id == null) {
	        throw new IllegalArgumentException("ID cannot be null.");
	    }
		
		Appointment foundAppointment = null;
		
		for (Appointment appointment : appointments) {
			if(appointment.getId().equals(id)) {
				foundAppointment = appointment;
				break;
			}
		}
		
		// A copy of the appointment is returned to allow for testing that compares different references of the appointment objects.
		Appointment copyAppointment = null;
		if (foundAppointment != null) {
			copyAppointment = new Appointment(foundAppointment);
		}
		
		return copyAppointment;
	}
	
	/**
	 * Gets a copy of the list of appointments, preventing changes to the internal appointment list.
	 * Returns an empty list if none exists.
	 * @return - the list of appointments or none if it is empty.
	 */
	public List<Appointment> getAll() {
	    List<Appointment> appointmentCopy = new ArrayList<>(appointments);
	    
	    return appointmentCopy;	    
	}
}

File: java-to-python-translation/appointment_service.py
"""Service layer for managing a collection of Appointment objects.

This module represents the Service Layer of the Service-Repository
layered architecture. AppointmentService is responsible for enforcing
collection-level business rules, ensuring the appointment list remains
valid, consistent, and well-organized. Field-level data integrity rules
such as length and format validation are handled separately in the
Appointment class, keeping each layer focused on a single responsibility.
"""

from appointment import Appointment

# NoSuchElementException is a custom exception that mirrors Java's
# NoSuchElementException behavior. Python has no direct built-in equivalent,
# so a custom class was created to preserve the same exception hierarchy
# and semantics as the original Java implementation during translation.
from exceptions import NoSuchElementException


class AppointmentService:
    """Manage a collection of Appointment objects following the Service Layer pattern.

    Provides add, delete, get, and get_all operations with full input
    validation before processing. The internal appointment list is private
    to ensure all modifications go through the service layer's validation
    logic. Direct list access would bypass duplicate ID checks and allow
    external code to insert unvalidated objects.

    AppointmentService intentionally does not include an edit method.
    Appointments are considered immutable once created. If changes are
    needed the appointment must be deleted and recreated. This reflects
    the original Java design decision and ensures appointment integrity
    is maintained throughout the lifecycle of each record.

    Copies are returned from get() and get_all() rather than stored
    references to prevent external code from modifying internal state
    directly and bypassing the service layer's business rules.
    """

    def __init__(self) -> None:
        """Initialize the AppointmentService with an empty appointment list.

        The internal _appointments list is private to enforce that all
        modifications go through the service layer's validation logic.
        Direct list access would bypass duplicate ID checks and allow
        insertion of invalid objects.
        """
        self._appointments = []

    def add(self, new_appointment: Appointment) -> None:
        """Add a new appointment to the service.

        Validates that the appointment is not None and that the ID is unique
        before adding. Duplicate ID prevention is a service layer responsibility
        since it is a collection-level business rule rather than a field-level
        data rule belonging in the Appointment class.

        Args:
            new_appointment: The Appointment object to add.

        Raises:
            ValueError: If the appointment is None or a duplicate ID exists.
        """
        if new_appointment is None:
            raise ValueError("Appointment cannot be null.")

        if self._find_appointment(new_appointment.get_id()) is not None:
            raise ValueError(f"Cannot add appointment with duplicate ID: {new_appointment.get_id()}")

        self._appointments.append(new_appointment)

    def delete(self, appointment_id: str) -> None:
        """Delete an appointment by its ID.

        Args:
            appointment_id: The ID of the appointment to delete.

        Raises:
            ValueError: If the ID is None.
            NoSuchElementException: If no appointment with the given ID exists.
        """
        if appointment_id is None:
            raise ValueError("ID cannot be null.")

        appointment_to_delete = self._find_appointment(appointment_id)

        if appointment_to_delete is None:
            raise NoSuchElementException(f"Appointment with ID {appointment_id} does not exist.")

        self._appointments.remove(appointment_to_delete)

    def get(self, appointment_id: str) -> Appointment | None:
        """Retrieve a copy of the appointment with the given ID.

        Returns a copy rather than the stored reference to prevent external
        code from obtaining a direct reference to the internal object and
        modifying it in ways that bypass the service layer's validation.

        Args:
            appointment_id: The ID of the appointment to retrieve.

        Returns:
            A copy of the matching Appointment, or None if not found.

        Raises:
            ValueError: If the ID is None.
        """
        if appointment_id is None:
            raise ValueError("ID cannot be null.")

        found_appointment = self._find_appointment(appointment_id)

        if found_appointment is not None:
            return Appointment.from_appointment(found_appointment)
        else:
            return None

    def get_all(self) -> list:
        """Return a list of copies of all appointments in the service.

        Copies are returned rather than stored references to prevent external
        code from modifying internal state directly. Returns an empty list
        if no appointments exist.

        Returns:
            A list of independent Appointment copies.
        """
        return [Appointment.from_appointment(appointment) for appointment in self._appointments]

    def _find_appointment(self, appointment_id: str) -> Appointment | None:
        """Search the internal list for an appointment matching the given ID.

        Private because it exposes a direct reference to the stored object
        rather than a copy. Only used internally by add(), delete(), and get()
        where direct reference access is needed for modification or removal.
        External callers should use get() which returns a safe copy instead.

        Args:
            appointment_id: The ID to search for.

        Returns:
            The matching Appointment reference, or None if not found.
        """
        found_appointment = None

        for appointment in self._appointments:
            if appointment.get_id() == appointment_id:
                found_appointment = appointment
                break

        return found_appointment